src/vi/mimihentai/src/eu/kanade/tachiyomi/extension/vi/mimihentai/MiMiHentai.kt

```diff
@@ -0,0 +1,201 @@
+package eu.kanade.tachiyomi.extension.vi.mimihentai
+
+import eu.kanade.tachiyomi.network.GET
+import eu.kanade.tachiyomi.network.asObservableSuccess
+import eu.kanade.tachiyomi.network.interceptor.rateLimit
+import eu.kanade.tachiyomi.source.model.Filter
+import eu.kanade.tachiyomi.source.model.FilterList
+import eu.kanade.tachiyomi.source.model.MangasPage
+import eu.kanade.tachiyomi.source.model.Page
+import eu.kanade.tachiyomi.source.model.SChapter
+import eu.kanade.tachiyomi.source.model.SManga
+import eu.kanade.tachiyomi.source.online.HttpSource
+import keiyoushi.utils.parseAs
+import keiyoushi.utils.tryParse
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import org.json.JSONObject
+import rx.Observable
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+class MiMiHentai : HttpSource() {
+    override val name: String = "MiMiHentai"
+
+    override val lang: String = "vi"
+
+    override val baseUrl: String = "https://mimihentai.com"
+
+    private val apiUrl: String = "$baseUrl/api/v1/manga"
+
+    override val supportsLatest: Boolean = true
+
+    override fun headersBuilder() = super.headersBuilder()
+        .add("Referer", "$baseUrl/")
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder()
+        .rateLimit(3)
+        .build()
+
+    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSS", Locale.US)
+
+    override fun latestUpdatesRequest(page: Int): Request = GET(
+        apiUrl.toHttpUrl().newBuilder().apply {
+            addPathSegments("tatcatruyen")
+            addQueryParameter("page", page.toString())
+            addQueryParameter("sort", "updated_at")
+        }.build(),
+        headers,
+    )
+
+    override fun latestUpdatesParse(response: Response): MangasPage {
+        val res = response.parseAs<MangaDTO>()
+        val manga = res.data.map {
+            SManga.create().apply {
+                title = it.title
+                thumbnail_url = it.coverUrl
+                setUrlWithoutDomain("$baseUrl/g/${it.id}")
```

Comment:
Extract these statements to a method in your `Manga` DTO (`toSManga()`). Then make the fields private. Same for chapters.

Use the `id` itself as the `url`, and then override in `getMangaUrl()` to make the WebView open the correct URL. Same for chapters.

Reply:
> Extract these statements to a method in your `Manga` DTO (`toSManga()`). Then make the fields private. Same for chapters.
> 
> Use the `id` itself as the `url`, and then override in `getMangaUrl()` to make the WebView open the correct URL. Same for chapters.

how can it be used with `getChapterUrl()` ? The chapter url has the following format 
```
chapterUrl = baseUrl.toHttpUrl().newBuilder().apply {
                addPathSegments(mangaDetailsParse(response).url)
                addPathSegments("chapter")
                addPathSegments(chapterDTO.title)
                addPathSegments("-")
                addPathSegments("${chapterDTO.id}")
            }.build().toString()
```
i don't know how to solve it